### PR TITLE
implements torch sdpa

### PR DESCRIPTION
**Pull request by LucasLLC**
Implements torch sdpa for mem_efficient kernel support

Using the mem_efficient kernel results in a ~15.5% faster training time per batch, going from a ~154ms/batch baseline to ~130ms/batch.

Potentially improves on overflow protection, since `mem_efficient` kernel scales q & k matrices before multiplying

Setting dropout to 0.0 is necessary to support the flash attention kernel. This will no longer be necessary after https://github.com/pytorch/pytorch/pull/92917 is landed

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -49,11 +48,16 @@ def forward(self, x):
...
-        att = F.softmax(att, dim=-1)
-        att = self.attn_dropout(att)
+        # the output of sdp = (batch, num_heads, seq_len, head_dim)
+        y, _ = torch.nn.functional._scaled_dot_product_attention(
+            q,
+            k,
+            v,
+            attn_mask=None,
+            dropout_p=self.dropout,
+            need_attn_weights=False,
```
Comment (drisspg) [inline model.py]: need_weights is removed on latest nightlies

and only returns the output so no need for `,_`

File: model.py [hunk 2]
```diff
@@ -49,11 +48,16 @@ def forward(self, x):
         q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
         v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
 
-        # causal self-attention; Self-attend: (B, nh, T, hs) x (B, nh, hs, T) -> (B, nh, T, T)
-        att = (q @ k.transpose(-2, -1)) * (1.0 / math.sqrt(k.size(-1)))
-        att = att.masked_fill(self.bias[:,:,:T,:T] == 0, float('-inf'))
-        att = F.softmax(att, dim=-1)
-        att = self.attn_dropout(att)
+        # the output of sdp = (batch, num_heads, seq_len, head_dim)
+        y, _ = torch.nn.functional._scaled_dot_product_attention(
```
Comment (drisspg) [inline model.py]: use the public function

File: model.py [hunk 3]
```diff
@@ -97,7 +101,7 @@ class GPTConfig:
     n_layer: int = 12
     n_head: int = 12
     n_embd: int = 768
-    dropout: float = 0.1
+    dropout: float = 0.0
```
Comment (drisspg) [inline model.py:98]: Leave a comment saying that this will not be needed after this PR:
https://github.com/pytorch/pytorch/pull/92917 lands

**Review by drisspg — CHANGES_REQUESTED** [review]
couple changes

File: model.py [hunk 4]
```diff
@@ -49,11 +48,15 @@ def forward(self, x):
         q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
         v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
 
-        # causal self-attention; Self-attend: (B, nh, T, hs) x (B, nh, hs, T) -> (B, nh, T, T)
-        att = (q @ k.transpose(-2, -1)) * (1.0 / math.sqrt(k.size(-1)))
-        att = att.masked_fill(self.bias[:,:,:T,:T] == 0, float('-inf'))
-        att = F.softmax(att, dim=-1)
-        att = self.attn_dropout(att)
+        # the output of sdp = (batch, num_heads, seq_len, head_dim)
+        y = torch.nn.functional.scaled_dot_product_attention(
```
Comment (drisspg) [inline model.py]: super nit: no need to change but
`y = F.scaled_dot_product_attention(q, k, v, None, self.dropout_p, is_causal=True)` looks alot for economically and could be a little more enticing?

**Review by drisspg — APPROVED** [review]

File: model.py (hunk 4 above)
Comment (LucasLLC) [reply model.py]: agreed!
